webserver/views/user.py

```diff
@@ -98,6 +99,63 @@ def import_data():
                            lastfm_username=lastfm_username)
 
 
+@user_bp.route("/upload", methods=['POST'])
+@login_required
+def upload():
+    if request.method == 'POST':
+        f = request.files['file']
+        if f:
+            augmented_listens = []
+            msb_listens = []
+            try:
+                jsonlist = json.load(f)
+                if not isinstance(jsonlist, list):
+                    raise ValueError
+            except ValueError:
+                raise BadRequest("Not a valid lastfmbackup file.")
+
+            for listen in jsonlist:
+                dic = _convert_to_native_format(listen)
+                _validate_listen(dic)
```

Comment:
This code is copied from the `submit` method in the API view. We shouldn't be duplicating this code. Instead, it should be put into a utilities module, which upload and submit can both call. In addition, making it a utility method makes it much easier to test.